azure/services/publicips/client.go

```diff
@@ -61,36 +68,134 @@ func (ac *AzureClient) Get(ctx context.Context, resourceGroupName, ipName string
 	return ac.publicips.Get(ctx, resourceGroupName, ipName, "")
 }
 
-// CreateOrUpdate creates or updates a static or dynamic public IP address.
-func (ac *AzureClient) CreateOrUpdate(ctx context.Context, resourceGroupName string, ipName string, ip network.PublicIPAddress) error {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "publicips.AzureClient.CreateOrUpdate")
+// CreateOrUpdateAsync creates or updates a static public IP in the specified resource group asynchronously.
+// It sends a PUT request to Azure and if accepted without error, the func will return a Future which can be used to track the ongoing
+// progress of the operation.
+func (ac *AzureClient) CreateOrUpdateAsync(ctx context.Context, spec azure.ResourceSpecGetter) (interface{}, azureautorest.FutureAPI, error) {
+	ctx, _, done := tele.StartSpanWithLogger(ctx, "publicips.AzureClient.CreateOrUpdateAsync")
 	defer done()
 
-	future, err := ac.publicips.CreateOrUpdate(ctx, resourceGroupName, ipName, ip)
+	ip, err := ac.publicIPParams(ctx, spec)
 	if err != nil {
-		return err
+		return nil, nil, errors.Wrapf(err, "failed to get desired parameters for group %s", spec.ResourceName())
+	} else if ip == nil {
+		// nothing to do here
+		return nil, nil, nil
 	}
+
+	future, err := ac.publicips.CreateOrUpdate(ctx, spec.ResourceGroupName(), spec.ResourceName(), *ip)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, reconciler.DefaultAzureCallTimeout)
+	defer cancel()
+
 	err = future.WaitForCompletionRef(ctx, ac.publicips.Client)
 	if err != nil {
-		return err
+		// if an error occurs, return the future.
+		// this means the long-running operation didn't finish in the specified timeout.
+		return nil, &future, err
 	}
-	_, err = future.Result(ac.publicips)
-	return err
+
+	result, err := future.Result(ac.publicips)
+	return result, nil, err
+}
+
+func (ac *AzureClient) publicIPParams(ctx context.Context, spec azure.ResourceSpecGetter) (*network.PublicIPAddress, error) {
+	var params interface{}
+	var existing interface{}
+
+	if existingPublicIP, err := ac.Get(ctx, spec.ResourceGroupName(), spec.ResourceName()); err != nil && !azure.ResourceNotFound(err) {
```

Comment:
this logic was extracted to async.go in #1874 so you should be able to get rid of it

Reply:
see https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1874/files#diff-0ab00cea0e2a73f9de32a12018949438868472c24c88553947106f4225eb0eb5L81 for an example of how to rebase